### PR TITLE
Improved spawn location calculations.

### PR DESCRIPTION
**Pull request by samoatesgames**
Note: Sorry for messing up the other PR. I rebased my fork, pushed and it auto-closed the other PR :(

 - Supports Overworld and Nether spawns.
 - Supports spawning under objects, but still above ground (e.g. under the leaves of a tree).
 - Protects against spawning in oceans.
 - Protects against spawning in water.
 - Uses a radial search about the origin, rather than a linear.
 - Correctly calculates Nether spawn on spawn world generation (fixes: https://github.com/cuberite/cuberite/issues/2548)
 - Fixes a bug in CheckPlayerSpawnPoint() where the X offset was used in both the X and Z coordinates (BLOCKTYPE BlockType = GetBlock(a_PosX + Coords[i].x, a_PosY, a_PosZ + Coords[i].x);)

### CODE REVIEW

File: src/World.cpp [hunk 1]
```diff
@@ -572,50 +572,158 @@ void cWorld::Start(void)
...
+
+		// Early out - Is the floor block air
+		if (FloorBlock == E_BLOCK_AIR)
+		{
+			continue;
+		}
+
+		// Is the floor block ok
+		bool ValidSpawnBlock = false;
+		for (int BlockIndex = 0; BlockIndex < ValidSpawnBlocksCount && !ValidSpawnBlock; ++BlockIndex)
```
Comment (worktycho) [inline src/World.cpp]: This won't unroll unless you remove the && !ValidSpawnBlock condition

**Comment by SphinxC0re** [comment]
There seems to be a style error somewhere

**Comment by samoatesgames** [comment]
@SphinxC0re haha around the bit that @worktycho  suggests removing :D Two birds, One push.

File: src/World.cpp [hunk 2]
```diff
@@ -325,10 +325,17 @@ void cWorld::SetNextBlockTick(int a_BlockX, int a_BlockY, int a_BlockZ)
 
 void cWorld::InitializeSpawn(void)
 {
+	// For the debugging builds, don't make the server build too much world upon start:
+#if defined(_DEBUG) || defined(ANDROID_NDK)
+	const int DefaultViewDist = 9;
+#else
+	const int DefaultViewDist = 20;  // Always prepare an area 20 chunks across, no matter what the actual cClientHandle::VIEWDISTANCE is
+#endif  // _DEBUG
```
Comment (madmaxoft) [inline src/World.cpp]: Style: Preprocessor directives should be indented like normal code

Comment (samoatesgames) [inline src/World.cpp]: Gah that's visual studio auto indenting when I moved. Damn you Microsoft!!!

File: src/World.cpp [hunk 3]
```diff
@@ -572,50 +572,158 @@ void cWorld::Start(void)
...
+		E_BLOCK_GRASS,
+		E_BLOCK_SAND,
+		E_BLOCK_SNOW,
+		E_BLOCK_SNOW_BLOCK,
+		E_BLOCK_NETHERRACK
+	};
+
+	static const int ValidSpawnBlocksCount = ARRAYCOUNT(ValidSpawnBlocks);
+
+	static const int HighestSpawnPoint = std::min((cChunkDef::Height >> 1) - 1, GetHeight(static_cast<int>(a_X), static_cast<int>(a_Z) + 16));
```
Comment (madmaxoft) [inline src/World.cpp]: Prefer to use arithmetic operations instead of bit-shifts.

File: src/World.cpp [hunk 4]
```diff
@@ -572,50 +572,158 @@ void cWorld::Start(void)
...
+
+	m_SpawnY = GetHeight(static_cast<int>(m_SpawnX), static_cast<int>(m_SpawnZ));
+	LOGWARNING("Did not find an acceptable spawnpoint. Generated a random spawnpoint position at {%.2f, %.2f, %.2f}", m_SpawnX, m_SpawnY, m_SpawnZ);
+}
+
+
+
+
+
+bool cWorld::CanSpawnAt(double a_X, double a_Z, double &a_Y)
```
Comment (madmaxoft) [inline src/World.cpp]: Style: Space after `&`

File: src/World.h [hunk 5]
```diff
@@ -1042,7 +1042,10 @@ class cWorld :
 	void UpdateSkyDarkness(void);
 
 	/** Generates a random spawnpoint on solid land by walking chunks and finding their biomes */
-	void GenerateRandomSpawn(void);
+	void GenerateRandomSpawn(int a_MaxSpawnRadius);
+
+	/** Can the specified coordinates be used as a spawn point? */
+	bool CanSpawnAt(double a_X, double a_Z, double &a_Y);
```
Comment (madmaxoft) [inline src/World.h]: All our functions use X, Y, Z, this one should, too.
Document the Y being an out-param.

File: src/World.cpp [hunk 6]
```diff
@@ -572,50 +572,158 @@ void cWorld::Start(void)
...
-			(biome != biOcean) && (biome != biFrozenOcean) &&  // The biome is acceptable (don't want a small ocean island)
-			!IsBlockWaterOrIce(GetBlock(SpawnX, GetHeight(SpawnX, SpawnZ), SpawnZ))  // The terrain is acceptable (don't want to spawn inside a lake / river)
-		)
+		// Found a usable biome
+		// Spawn chunks so we can find a nice spawn.
+		int ChunkX = 0, ChunkZ = 0;
+		cChunkDef::BlockToChunk(BiomeOffset.x, BiomeOffset.z, ChunkX, ChunkZ);
+		cSpawnPrepare::PrepareChunks(*this, ChunkX, ChunkZ, a_MaxSpawnRadius);
+		break;
+	} while ((--BiomeCheckCount) > 0);
```
Comment (madmaxoft) [inline src/World.cpp]: Prefer a regular `for` loop

File: src/World.cpp [hunk 7]
```diff
@@ -572,50 +572,158 @@ void cWorld::Start(void)
...
+		Vector3i(HalfChunk, 0, -HalfChunk),
+		Vector3i(0, 0, -HalfChunk),
+		Vector3i(-HalfChunk, 0, -HalfChunk),
+		Vector3i(-HalfChunk, 0, 0),
+		Vector3i(-HalfChunk, 0, HalfChunk),
+	};
+
+	static const int PerRadiSearchCount = ARRAYCOUNT(ChunkOffset);
+
+	for (int RadiusOffset = 1; RadiusOffset < (a_MaxSpawnRadius << 1); ++RadiusOffset)
```
Comment (madmaxoft) [inline src/World.cpp]: Prefer arithmetic to bitshift operations

**Comment by madmaxoft** [comment]
I have an issue with the new search code. It searches near chunks first, then extends further and further away. However, suppose the initial choice is in an ocean. The general assumption is that the nearby chunks will all be ocean, rather than land, so it makes sense to go as far away as possible as fast as possible. That's why the current search code actually walks away from the initial choice. I think that's a much better approach than generating lots of needless ocean chunks.

**Comment by samoatesgames** [comment]
It has a check for ocean biomes and first moves out of them.
This means the search origin is in a valid biome, searching around that is less likely to move into another invalid biome, where as the linear search could easily move into a different ocean biome.

**Comment by SphinxC0re** [comment]
But what about custom generators? Do they have to bring their own spawngenerator?

**Comment by samoatesgames** [comment]
The case of custom generators would have to be supported by having a config option for invalid spawn biomes.

The current linear search would still have issues with custom generators.

**Comment by madmaxoft** [comment]
Ah, I've missed that. It really does try to get out of ocean fast.

@SphinxC0re custom generators such as in plugin-provided generators? Those haven't worked with the old code either and they don't work with the new code, but there's currently no real way to get them to cooperate. We'd need to break the generator-to-plugin bindings into a bit smaller facets in order to achieve that.

**Comment by samoatesgames** [comment]
All style issues should have been corrected.

**Comment by samoatesgames** [comment]
Are there any other improvements people want made?
Or are people generally not happy with the algorithm used?
:)

**Comment by SphinxC0re** [comment]
Thank you so much for the change @samoatesgames

**Pull request merged**